In [1]:
link = "ParisHousingClass99.88.csv"

Sources :
Part of codes :
https://www.javatpoint.com/machine-learning-decision-tree-classification-algorithm

https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html

The data file I have attached as Paring Housing file

Data Source :
https://www.kaggle.com/mssmartypants/paris-housing-classification?select=ParisHousingClass99.88.csv

This dataset is to train the features of numberOfRooms 	hasYard 	hasPool 	made 	isNewBuilt 	hasGuestRoom 	price of data

to predict either the paris house is luxury or basic

In [1]:
# importing libraries  
import numpy as nm  
import matplotlib.pyplot as mtp  
import pandas as pd  
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.preprocessing import LabelEncoder
#LabelEncoder is to convert string data to numeric data

In [3]:
#importing datasets  
data_set= pd.read_csv(link) 
data_set


squareMeters  numberOfRooms  hasYard  hasPool  floors  cityCode  \
0            75523              3        0        1      63      9373   
1            80771             39        1        1      98     39381   
2            55712             58        0        1      19     34457   
3            32316             47        0        0       6     27939   
4            70429             19        1        1      90     38045   
...            ...            ...      ...      ...     ...       ...   
9995          1726             89        0        1       5     73133   
9996         44403             29        1        1      12     34606   
9997         83841              3        0        0      69     80933   
9998         59036             70        0        0      96     55856   
9999          1440             84        0        0      49     18412   

      cityPartRange  numPrevOwners  made  isNewBuilt  hasStormProtector  \
0                 3              8  2005           0                  1   
1                 8              6  2015           1                  0   
2                 6              8  2021           0                  0   
3                10              4  2012           0                  1   
4                 3              7  1990           1                  0   
...             ...            ...   ...         ...                ...   
9995              7              6  2009           0                  1   
9996              9              4  1990           0                  1   
9997             10             10  2005           1                  1   
9998              1              3  2010           0                  1   
9999              6             10  1994           1                  0   

      basement  attic  garage  hasStorageRoom  hasGuestRoom      price  \
0         4313   9005     956               0             7  7559081.5   
1         3653   2436     128               1             2  8085989.5   
2         2937   8852     135               1             9  5574642.1   
3          659   7141     359               0             3  3232561.2   
4         8435   2429     292               1             4  7055052.0   
...        ...    ...     ...             ...           ...        ...   
9995      9311   1698     218               0             4   176425.9   
9996      9061   1742     230               0             0  4448474.0   
9997      8304   7730     345               1             9  8390030.5   
9998      2590   6174     339               1             4  5905107.0   
9999      8485   2024     278               1             6   146708.4   

     category  
0       Basic  
1      Luxury  
2       Basic  
3      Luxury  
4      Luxury  
...       ...  
9995    Basic  
9996    Basic  
9997    Basic  
9998    Basic  
9999    Basic  

[10000 rows x 18 columns]

In [4]:
#convert string in int
column_list = data_set.columns

cvt = LabelEncoder()

for i in column_list:
    if data_set[i].dtype != 'int32':
        data_set[i] = cvt.fit_transform(data_set[i])
        
#remove("category")
feature = column_list[:-1]
feature


Index(['squareMeters', 'numberOfRooms', 'hasYard', 'hasPool', 'floors',
       'cityCode', 'cityPartRange', 'numPrevOwners', 'made', 'isNewBuilt',
       'hasStormProtector', 'basement', 'attic', 'garage', 'hasStorageRoom',
       'hasGuestRoom', 'price'],
      dtype='object')

In [5]:
print(data_set.iloc[:,:-1])

      squareMeters  numberOfRooms  hasYard  hasPool  floors  cityCode  \
0             7178              2        0        1      62       917   
1             7680             38        1        1      97      3699   
2             5287             57        0        1      18      3284   
3             3038             46        0        0       5      2676   
4             6707             18        1        1      89      3579   
...            ...            ...      ...      ...     ...       ...   
9995           182             88        0        1       4      6867   
9996          4205             28        1        1      11      3306   
9997          7976              2        0        0      68      7641   
9998          5603             69        0        0      95      5249   
9999           154             83        0        0      48      1769   

      cityPartRange  numPrevOwners  made  isNewBuilt  hasStormProtector  \
0                 2              7    15        

In [6]:
feature_col = list(range(0,data_set.shape[1]-1))
feature_col

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]

In [7]:
#self implementation
def display_data(data, fea_col):
    return data_set.iloc[:,fea_col]
features = display_data(data_set, feature_col)

print("The features are : \n")
features
    
    

The features are : 



squareMeters  numberOfRooms  hasYard  hasPool  floors  cityCode  \
0             7178              2        0        1      62       917   
1             7680             38        1        1      97      3699   
2             5287             57        0        1      18      3284   
3             3038             46        0        0       5      2676   
4             6707             18        1        1      89      3579   
...            ...            ...      ...      ...     ...       ...   
9995           182             88        0        1       4      6867   
9996          4205             28        1        1      11      3306   
9997          7976              2        0        0      68      7641   
9998          5603             69        0        0      95      5249   
9999           154             83        0        0      48      1769   

      cityPartRange  numPrevOwners  made  isNewBuilt  hasStormProtector  \
0                 2              7    15           0                  1   
1                 7              5    25           1                  0   
2                 5              7    31           0                  0   
3                 9              3    22           0                  1   
4                 2              6     0           1                  0   
...             ...            ...   ...         ...                ...   
9995              6              5    19           0                  1   
9996              8              3     0           0                  1   
9997              9              9    15           1                  1   
9998              0              2    20           0                  1   
9999              5              9     4           1                  0   

      basement  attic  garage  hasStorageRoom  hasGuestRoom  price  
0         2731   5638     856               0             7   7580  
1         2314   1509      28               1             2   8106  
2         1860   5536      35               1             9   5584  
3          442   4447     259               0             3   3196  
4         5366   1505     192               1             4   7092  
...        ...    ...     ...             ...           ...    ...  
9995      5913   1073     118               0             4    195  
9996      5762   1100     130               0             0   4439  
9997      5282   4825     245               1             9   8417  
9998      1648   3842     239               1             4   5914  
9999      5395   1273     178               1             6    161  

[10000 rows x 17 columns]

In [8]:
#feature size
features.shape

(10000, 17)

In [9]:

label_index = data_set.shape[1]-1
label = display_data(data_set,label_index )

print("The label column is : \n")
label

The label column is : 



0       0
1       1
2       0
3       1
4       1
       ..
9995    0
9996    0
9997    0
9998    0
9999    0
Name: category, Length: 10000, dtype: int32

In [10]:

  
#Extracting Independent and dependent Variable  
x= data_set.iloc[:,feature_col].values  

y= data_set.iloc[:, label_index].values  
  
# Splitting the dataset into training and test set.  
from sklearn.model_selection import train_test_split  
x_train, x_test, y_train, y_test= train_test_split(x, y, test_size= 0.2, random_state=0)  

print(x_train.shape,x_test.shape, y_train.shape,y_test.shape) 

#feature Scaling  
from sklearn.preprocessing import StandardScaler    
st_x= StandardScaler()  
x_train= st_x.fit_transform(x_train)    
x_test= st_x.transform(x_test) 



(8000, 17) (2000, 17) (8000,) (2000,)


In [11]:
x_train

array([[-0.98192119, -0.22088849,  0.98955455, ...,  0.98683663,
         1.26788717, -0.98197588],
       [-0.97718748,  0.78872832,  0.98955455, ..., -1.01333896,
        -0.61896668, -0.97714277],
       [ 1.50109024, -0.9868047 ,  0.98955455, ...,  0.98683663,
         0.95341153,  1.49982543],
       ...,
       [ 1.37546494,  0.51021334, -1.01055571, ..., -1.01333896,
        -0.30449104,  1.37278372],
       [ 1.5750089 ,  1.27612954, -1.01055571, ...,  0.98683663,
         0.0099846 ,  1.57197684],
       [-1.10608997,  0.16206961, -1.01055571, ..., -1.01333896,
         0.63893588, -1.10729148]])

In [12]:
x_test

array([[-1.23244353, -1.54383466, -1.01055571, ..., -1.01333896,
        -1.24791796, -1.23295231],
       [-1.53649318, -0.081631  , -1.01055571, ..., -1.01333896,
         0.0099846 , -1.53743815],
       [ 0.45129966, -0.39496036, -1.01055571, ...,  0.98683663,
         0.95341153,  0.45242174],
       ...,
       [ 0.42071263,  0.54502771,  0.98955455, ...,  0.98683663,
        -0.93344232,  0.42238742],
       [ 1.35252466,  1.45020141, -1.01055571, ...,  0.98683663,
        -1.5623936 ,  1.35241562],
       [ 1.57537303, -0.01200226,  0.98955455, ..., -1.01333896,
         0.32446024,  1.57266729]])

In [13]:
#this part I tested how each function works

In [14]:
#Fitting Decision Tree classifier to the training set  
from sklearn.tree import DecisionTreeClassifier 
#classifier= DecisionTreeClassifier(criterion='entropy', random_state=0)
classifier =DecisionTreeClassifier(criterion = "entropy", random_state = 100,max_depth = 3, min_samples_leaf = 5)
classifier.fit(x_train, y_train)  

DecisionTreeClassifier(criterion='entropy', max_depth=3, min_samples_leaf=5,
                       random_state=100)

In [15]:
#Predicting the test set result  
y_pred= classifier.predict(x_test)  
y_pred


array([0, 0, 0, ..., 0, 0, 1])

In [16]:
y_test


array([0, 0, 0, ..., 0, 0, 1])

# Below is the part I train my data in different models and put together model, predict, validate data, and compare with other models.

In [17]:
#Creating the Confusion matrix  
from sklearn.metrics import confusion_matrix  
cm= confusion_matrix(y_test, y_pred)  
print("Matric ",cm)

Matric  [[1463    0]
 [   4  533]]


In [18]:

def matric_translate(confuse_matrix):
    incorrect_pred = cm[0,1]+cm[1,0]
    correct_pred = cm[0,0]+cm[1,1]
    print("Incorrect prediction number is ",incorrect_pred)
    print("Correct prediction number is ", correct_pred)
    
matric_translate(cm)

Incorrect prediction number is  4
Correct prediction number is  1996


In [19]:
print ("Accuracy : ",accuracy_score(y_test,y_pred)*100)

Accuracy :  99.8


In [20]:
print("Report : ",classification_report(y_test, y_pred))

Report :                precision    recall  f1-score   support

           0       1.00      1.00      1.00      1463
           1       1.00      0.99      1.00       537

    accuracy                           1.00      2000
   macro avg       1.00      1.00      1.00      2000
weighted avg       1.00      1.00      1.00      2000



In [21]:
#self implemeantion these training functions
def train_with_entropy_generic():
    classifier= DecisionTreeClassifier(criterion='entropy', random_state=100)
    classifier.fit(x_train, y_train) 
    return classifier
    

In [22]:

def train_with_entropy_specific(m, min_l):
    classifier =DecisionTreeClassifier(criterion = "entropy", random_state = 100,max_depth=m, min_samples_leaf =min_l)
    classifier.fit(x_train, y_train) 
    return classifier

In [23]:
def train_with_gini_generic():
    classifier= DecisionTreeClassifier(criterion='gini', random_state=100)
    classifier.fit(x_train, y_train) 
    return classifier

In [24]:
def train_with_gini_specific(m, min_l):
    classifier =DecisionTreeClassifier(criterion = "gini", random_state = 100,max_depth=m , min_samples_leaf=min_l)
    classifier.fit(x_train, y_train) 
    return classifier

In [25]:
#self implementation
def matric_translate(confuse_matrix):
    incorrect_pred = cm[0,1]+cm[1,0]
    correct_pred = cm[0,0]+cm[1,1]
    print("Incorrect prediction number is ",incorrect_pred)
    print("Correct prediction number is ", correct_pred)
    

In [26]:
#self putting together
def run(model):

    
    #train generic entropy
    y_pred = model.predict(x_test)

    #Creating the Confusion matrix  

    cm= confusion_matrix(y_test, y_pred)  
    print("Confusion matrix ")
    print(cm)

    matric_translate(cm)
    print ("Accuracy : ",accuracy_score(y_test,y_pred)*100)
    print("Report : ",classification_report(y_test, y_pred))



In [27]:
print("\nModel 1 : generic entropy training \n********************************")
generic_entrop =train_with_entropy_generic()
run(generic_entrop)


Model 1 : generic entropy training 
********************************
Confusion matrix 
[[1459    4]
 [   4  533]]
Incorrect prediction number is  4
Correct prediction number is  1996
Accuracy :  99.6
Report :                precision    recall  f1-score   support

           0       1.00      1.00      1.00      1463
           1       0.99      0.99      0.99       537

    accuracy                           1.00      2000
   macro avg       0.99      0.99      0.99      2000
weighted avg       1.00      1.00      1.00      2000



In [28]:
print("\nModel 2 : specific entropy training with max depth is 5 and 3 as min of leaves.\n********************************")

specific_entrop = train_with_entropy_specific(5, 3)

run(specific_entrop)


Model 2 : specific entropy training with max depth is 5 and 3 as min of leaves.
********************************
Confusion matrix 
[[1463    0]
 [   4  533]]
Incorrect prediction number is  4
Correct prediction number is  1996
Accuracy :  99.8
Report :                precision    recall  f1-score   support

           0       1.00      1.00      1.00      1463
           1       1.00      0.99      1.00       537

    accuracy                           1.00      2000
   macro avg       1.00      1.00      1.00      2000
weighted avg       1.00      1.00      1.00      2000



In [29]:
print("\nModel 3 : specific entropy training with max depth is 7 and 3 as min of leaves.\n********************************")



specific_entrop2 = train_with_entropy_specific(7, 3)
run(specific_entrop2)


Model 3 : specific entropy training with max depth is 7 and 3 as min of leaves.
********************************
Confusion matrix 
[[1463    0]
 [   4  533]]
Incorrect prediction number is  4
Correct prediction number is  1996
Accuracy :  99.8
Report :                precision    recall  f1-score   support

           0       1.00      1.00      1.00      1463
           1       1.00      0.99      1.00       537

    accuracy                           1.00      2000
   macro avg       1.00      1.00      1.00      2000
weighted avg       1.00      1.00      1.00      2000



In [30]:
print("\nModel 4  : specific entropy training with max depth is 2 and 2 as min of leaves.\n********************************")



specific_entrop2 = train_with_entropy_specific(2, 2)
run(specific_entrop2)


Model 4  : specific entropy training with max depth is 2 and 2 as min of leaves.
********************************
Confusion matrix 
[[1463    0]
 [   4  533]]
Incorrect prediction number is  4
Correct prediction number is  1996
Accuracy :  99.8
Report :                precision    recall  f1-score   support

           0       1.00      1.00      1.00      1463
           1       1.00      0.99      1.00       537

    accuracy                           1.00      2000
   macro avg       1.00      1.00      1.00      2000
weighted avg       1.00      1.00      1.00      2000



In [31]:
print("\nModel 5 : generic gini training\n********************************")

generic_gini = train_with_gini_generic()

run(generic_gini)


Model 5 : generic gini training
********************************
Confusion matrix 
[[1463    0]
 [   4  533]]
Incorrect prediction number is  4
Correct prediction number is  1996
Accuracy :  99.8
Report :                precision    recall  f1-score   support

           0       1.00      1.00      1.00      1463
           1       1.00      0.99      1.00       537

    accuracy                           1.00      2000
   macro avg       1.00      1.00      1.00      2000
weighted avg       1.00      1.00      1.00      2000



In [32]:
print("\nModel 6 : specific gini training with 5 max depth and 3 min leaves\n********************************")

specific_gini = train_with_gini_specific(5,3)


run(specific_gini)


Model 6 : specific gini training with 5 max depth and 3 min leaves
********************************
Confusion matrix 
[[1463    0]
 [   4  533]]
Incorrect prediction number is  4
Correct prediction number is  1996
Accuracy :  99.8
Report :                precision    recall  f1-score   support

           0       1.00      1.00      1.00      1463
           1       1.00      0.99      1.00       537

    accuracy                           1.00      2000
   macro avg       1.00      1.00      1.00      2000
weighted avg       1.00      1.00      1.00      2000



In [33]:
print("\nModel 7 : specific gini training with 7 max depth and 3 min leaves\n********************************")

specific_gini2 = train_with_gini_specific(7,3)


run(specific_gini2)


Model 7 : specific gini training with 7 max depth and 3 min leaves
********************************
Confusion matrix 
[[1463    0]
 [   4  533]]
Incorrect prediction number is  4
Correct prediction number is  1996
Accuracy :  99.8
Report :                precision    recall  f1-score   support

           0       1.00      1.00      1.00      1463
           1       1.00      0.99      1.00       537

    accuracy                           1.00      2000
   macro avg       1.00      1.00      1.00      2000
weighted avg       1.00      1.00      1.00      2000



In [34]:
print("\nModel 8 : specific gini training with None max depth and 2 min leaves\n********************************")

specific_gini3 = train_with_entropy_specific(None,2)


run(specific_gini3)


Model 8 : specific gini training with None max depth and 2 min leaves
********************************
Confusion matrix 
[[1459    4]
 [   4  533]]
Incorrect prediction number is  4
Correct prediction number is  1996
Accuracy :  99.6
Report :                precision    recall  f1-score   support

           0       1.00      1.00      1.00      1463
           1       0.99      0.99      0.99       537

    accuracy                           1.00      2000
   macro avg       0.99      0.99      0.99      2000
weighted avg       1.00      1.00      1.00      2000



My observations :
Between my so called "generic" and "specific" training models, the specific performs better by 0.02%.
In general : generic has 99.6 accuracy while the specific got 99.8 percentage of accuracy.
In Specific : (5,3),(7,3) has 99.8 accuracy for both models.
In (None,2) has 99.6 accuracy on model 2 and (2,2) has 99.8 on model1.

My issue is to look for data that is mostly I found out this decision tree is perfect suit for classification 
more than regression problem.

I splited this data to train, test in 80:20.
I played around with these completed implementation online. 
However, later on, I decided to adjust to test with different training models. 
The hardest part is to find perfect data that has met requirement. Second of that is how to decide that features 
can be determined the prediction. 




Sources :
Part of codes :
https://www.javatpoint.com/machine-learning-decision-tree-classification-algorithm

https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html

The data file I have attached as Paring Housing file

Data Source :
https://www.kaggle.com/mssmartypants/paris-housing-classification?select=ParisHousingClass99.88.csv